### PR TITLE
BFD-2915: Employ Cipher

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2915](https://jira.cms.gov/browse/BFD-2915)

**User Story or Bug Summary:**

As a BFD Ops Engineer I want our sensitive configuration strategy to be simplified so that I can get on with my life.

---

### What Does This PR Do?

This PR updates our configuration management scheme to use `cipher` instead of Ansible Vault. The changes made are intended to have no effect on downstream consumers of SSM configuration, so no parameter values or names have changed as part of this PR.

In more detail, this PR:

- Merges `.eyaml` and `.yaml` configuration files into a single file per-environment allowing plaintext and secret values to live in the same place
- Changes our `.yaml` configuration to be truly hierarchical using nested keys and objects
- Changes the yaml helper scripts that ingest the aforementioned configuration files to:
  - Automatically convert a nested key into a pathname that SSM understands (i.e. an unencrypted value referenced by `bfd.test.common.value` becomes `/bfd/test/common/nonsensitive/value`)
  - Automatically infer the sensitivity of a given parameter by whether or not it's encrypted via `cipher` and include that sensitivity in the aforementioned, generated SSM pathname following existing conventions
- Fixes a small bug in `cipher` that was causing issues with using an `EDITOR` value that specifies a flag (i.e. `EDITOR=code -w`)
- Updates the `bfd-cbc-build` build process to include `kotlin` and to also "build" the `cipher` Kotlin script so that its dependencies are cached for use in our pipelines
- Updates the pre-commit hooks to check for various sentinel values indicating unencrypted secrets and to block any commits that might expose those secrets
- Updates all Jenkins pipelines using `bfd-cbc-build` to use the latest version including Kotlin tagged as `public.ecr.aws/c2o1d8s9/bfd-cbc-build:jdk17-mvn3-tfenv3-kt1.9-latest`
- Updates various `README`s

This PR has been verified by:

- `terraform plan`ning `mgmt`'s `base_config`, verifying that there are no changes to our existing SSM parameters
- `terraform plan`ning `base` in all established environmentsi, verifiying that there are no unexpected changes to our existing SSM parameters
- building the updated `bfd-cbc-build` Docker image in Jenkins, verifying that the image builds successfully
- running a special, one-off Jenkins pipeline using the new `bfd-cbc-build` Docker container that just `terraform plan`s `base` in `test`, verifying that this pipeline was able to `plan` successfully using the new Docker container and that there were no unexpected changes

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [ ] Yes
    - [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- N/A

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by brandoncruz3 — DISMISSED** [review]
🚀

File: ops/terraform/services/base/values/ephemeral.yaml [hunk 1]
```diff
@@ -1,78 +1,73 @@
...
-/bfd/${env}/server/nonsensitive/lb_is_public: false
-/bfd/${env}/server/nonsensitive/lb_ingress_port: 443
-/bfd/${env}/server/nonsensitive/lb_egress_port: 7443
-# NOTE: lb_vpc_peerings_json stores a json string that will be decoded within the server
-# Terraservice into a list. The ephemeral variant is purposefully empty to ensure there are no
-# peerings to ephemeral environments
-/bfd/${env}/server/nonsensitive/lb_vpc_peerings_json: >-
-  []
+bfd: 
+  ${env}:
```
Comment (mackec1) [inline ops/terraform/services/base/values/ephemeral.yaml:3]: Really like how this is now organized!

**Review by mackec1 — APPROVED** [review]
Looks good...there was one shellcheck hiccup but the file is not part of your changeset....

**Review by mjburling — CHANGES_REQUESTED** [review]
@malessi I can't say enough about this. Once approved, I think it improves maintainability, usability, and with the MRK support, I feel this is a more reliable solution. Not to mention the additional key control we get by leveraging AWS IAM. Thanks for putting this together!

Ahead of acceptance, I think there are some minor issues that I think we'd be remiss not to address now.

* I'd like the helper scripts to be a _little_ more helpful. I don't think a user who needs to view/edit the values files should be forced to identify the specific Key ID for the encrypted values. Our current implementation allows us to use the environment name to inform the context and support a sensible default. Users should only need to identify the  _seed environment_, e.g. `scripts/edit-yaml.sh test`. For brevity's sake, I think the following snippet should suffice for demonstration purposes:
```bash
# ..snipped..
SCRIPT_DIR="$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")"
readonly SCRIPT_DIR

MODULE_DIR="$(dirname "$SCRIPT_DIR")"
readonly MODULE_DIR

REPO_ROOT="$(cd "$SCRIPT_DIR" && git rev-parse --show-toplevel)"
readonly REPO_ROOT

BFD_SEED_ENV="$1"
readonly BFD_SEED_ENV

YAML_FILE="${MODULE_DIR}/values/${BFD_SEED_ENV}.yaml"
readonly YAML_FILE

CMK="$2"

CMK_LOOKUP="$(aws kms describe-key --key-id "alias/bfd-${BFD_SEED_ENV}-config-cmk" --query KeyMetadata.Arn --output text)"
readonly CMK_LOOKUP

CMK_ARN="${CMK:="$CMK_LOOKUP"}"
readonly CMK_ARN
# ..snipped..
```

* there are a handful of places in the base (maybe the base_config, too?) module README that need some minor updates to help familiarize newcomers. For instance, "eyaml" is still referenced in places like examples that invite users to run `scripts/read-and-decrypt-eyaml.sh prod-sbx`.

File: ops/terraform/env/mgmt/base_config/scripts/edit-yaml.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+CMK_LOOKUP="$(
+  # Only attempt to lookup the CMK if the override is undefined
+  [[ -z "$CMK_ARN_OVERRIDE" ]] &&
+    aws kms describe-key \
+      --key-id "alias/bfd-${BFD_SEED_ENV}-config-cmk" \
+      --query KeyMetadata.Arn \
+      --output text
+)"
+readonly CMK_LOOKUP
```
Comment (mjburling) [inline ops/terraform/env/mgmt/base_config/scripts/edit-yaml.sh:37]: 👍

**Review by mjburling — APPROVED** [review]
🥇

**Pull request merged**